libjulia/optimiser/UnusedPruner.cpp

```diff
@@ -0,0 +1,107 @@
+/*(
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimisation stage that removes unused variables and functions.
+ */
+
+#include <libjulia/optimiser/UnusedPruner.h>
+
+#include <libjulia/optimiser/NameCollector.h>
+#include <libjulia/optimiser/Semantics.h>
+#include <libjulia/optimiser/Utilities.h>
+
+#include <libsolidity/inlineasm/AsmData.h>
+
+#include <boost/algorithm/cxx11/none_of.hpp>
+
+using namespace std;
+using namespace dev;
+using namespace dev::julia;
+
+UnusedPruner::UnusedPruner(Block& _ast)
+{
+	ReferencesCounter counter;
+	counter(_ast);
+
+	m_references = counter.references();
+}
+
+void UnusedPruner::operator()(Block& _block)
+{
+	for (auto&& statement: _block.statements)
+		if (statement.type() == typeid(FunctionDefinition))
+		{
+			FunctionDefinition& funDef = boost::get<FunctionDefinition>(statement);
+			if (!used(funDef.name))
+			{
+				subtractReferences(ReferencesCounter::countReferences(funDef.body));
+				statement = Block{std::move(funDef.location), {}};
+			}
+		}
+		else if (statement.type() == typeid(VariableDeclaration))
+		{
+			VariableDeclaration& varDecl = boost::get<VariableDeclaration>(statement);
+			if (boost::algorithm::none_of(
+				varDecl.variables,
+				[=](TypedName const& _typedName) { return used(_typedName.name); }
```

Comment:
Too bad because of tuple assignment with variable declarations we couldn't just remove variables which are not used. Perhaps leave that as a comment so nobody tries to "fix" it.